### ISSUE TITLE
api: create documentation

### ISSUE THREAD

**Issue by jorgeorpinel** [iterative/dvc.org#463]
> See src @ https://github.com/iterative/dvc/blob/master/dvc/api.py

`dvc.api.get_url`
`dvc.api.open`
`dvc.api.read`

~~Please more insights here~~ More details in https://github.com/iterative/dvc.org/issues/463#issuecomment-514139792.

Also, please update the one mention to API in the data registry (which will me merged with #818) per https://github.com/iterative/dvc.org/pull/818#issuecomment-565083934.

UPDATE:

- [ ] Also (at least basic docs for) `dvc.api.summon` and open another issue to complete it later.
- [ ] And try to do app: automatic links for API function names in md #910 once #908 is merged.

**Comment by shcheklein** [comment]
@Suor can you describe it briefly, please :)

**Comment by Naba7** [comment]
@jorgeorpinel @shcheklein can you please explain this a bit more. I would like to take this up.

**Comment by shcheklein** [comment]
@Suor ? can you give a summary or is there a link? we should probably put docstrings around APIs before we release it.

**Comment by Naba7** [comment]
sorry, I am unable to search ``dvc.api.get_content`` . Do you mean ``get_url`` or ``__getattr__`` ?

**Comment by shcheklein** [comment]
@Naba7 it's `read` now, I believe. There are only three public methods in `dvc.api`. Those should be described. And let's create a ticket on iterative/dvc to update docstrings for those APIs.

**Comment by Suor** [comment]
They have some short docstrings, I will update them based on future docs or discussion here if we decide to do that.

So, what is this about? There are three public things in `dvc.api` now:
1. `read(path, repo=None, rev=None, remote=None, mode="r", encoding=None)` - returns the contents of an artifact as a bytes object or a string.
2. `get_url(path, repo=None, rev=None, remote=None)` -  returns an url of an artifact.
3. `open(path, repo=None, rev=None, remote=None, mode="r", encoding=None)` - opens an artifact as a file, may only be used as context manager:
    ```python
    with dvc.api.open("path/to/data.csv", remote="my-s3", encoding="utf-8") as f:
        for line in f:
            process(line)
    ```

Arguments always mean the same:

`path` - a path to an artifact, relative to repo root,
`repo` - a path or git url of a repo,
`rev` - revision, i.e. a branch, a tag, a sha. This only works with an url in repo,
`remote` - a name of a remote to fetch artifact from/give url to
`mode` - a mode with which we open a file, the only sensible options are `r`/`rt` and `rb` 
`encoding` - an encoding used to decode contents to a string

`mode` and `encoding` mirror their namesakes builtin `open()` has.

**Comment by Naba7** [comment]
This issue is meant for improving docstrings around ``dvc.api`` [Link](https://github.com/iterative/dvc/issues/2316)

**Comment by shcheklein** [comment]
k, thanks @Suor . @Naba7 now we need to come with a good place and a format for it. Probably, we need a separate top-level section. `API reference` similar to command reference we have.

**Comment by Naba7** [comment]
Shall we describe ``read``, ``open``, and ``get_url`` in different sections inside ``API Reference``?
We can have another section for describing ``path``, ``repo``, ``rev``, ``remote``, ``mode``, ``encoding``.

**Comment by shcheklein** [comment]
@Naba7 yes, it should be a separate section per call. At least for now.

`path, repo, rev, remote, mode, encoding` - are not APIs. `repo` is, but it's not officially released yet.

Can you take a look for some good projects that have APIs documented to come up with a good page template for this?

**Comment by Naba7** [comment]
How about this format [angularJS](https://docs.angularjs.org/api) or [microsoft](https://docs.microsoft.com/en-us/graph/api/overview?view=graph-rest-1.0) or [Azure](https://docs.microsoft.com/en-us/rest/api/azure/)?

We can use these formats and write a introductory page listing all the APIs and linking them to GitHub

**Comment by jorgeorpinel** [comment]
@Suor are `repo=None, rev=None, remote=None` default values actually those? Or do they get turned into `repo='.', rev='HEAD', remote=(read from config file)`? Probably important to document (both in docstring and) in the API ref.

@shcheklein re
> Probably, we need a separate top-level section. `API reference` similar to command reference we have.

Agree, perhaps in docs path `/api` and the index page for that section could explain what is the API and how to start using it. Actually its not that obvious! I'm not 100% sure what we mean by "the DVC API" for example. Is it a Python library people can install separately?

**Comment by shcheklein** [comment]
that's right. I would say that it's not separate though, it's the same DVC package.

**Comment by Naba7** [comment]
Sorry, I didn't understand it. I thought DVC-API is same as DVC package being installed by binary-files or by any other means and it is a separate package to run on top of git and we can  implement it locally or on remote machines.

**Comment by shcheklein** [comment]
@Naba7 ?? I'm not sure I understand this. DVC api is just a python module defined in the DVC itself. No need to install any extras to being able to use it.

**Comment by jorgeorpinel** [comment]
Suggested intro. text for the new `/api-reference` section index page:

The DVC API is part of the `dvc` Python module installed along with DVC. You may include it in your Python code with the following line:

```python
from dvc import api as dvcapi
```

Then list `read`, `open`, and `get_url` in a bullet list, linked to their own pages (which should also have left pane navigation items).

**Comment by Naba7** [comment]
@shcheklein Okay. I get it now. I thought for getting DVC-api we need to download it separately.
@jorgeorpinel As @shcheklein had said that ``repo`` is also an api, we should include it in the intro.
And
1) ... some of the core functions of DVC such as ``add, push, pull, commit, checkout, etc.,`` ...
 
2) Writing one-liner for ``read,open, get_url, repo``(?) such as:
- read[link] - returns the contents of an artifact as a bytes object or a string.
- get_url[link] - returns an url of an artifact.
- open[link] - opens an artifact as a file.
- repo[link] - contains all the core command-files that are used in Python scripts.

3) Other followed up pages may contain vivid description and followed up by examples. What more can we include here?

**Comment by jorgeorpinel** [comment]
I think that's good enough to start a PR. Please let us know, thanks!

**Comment by Suor** [comment]
@jorgeorpinel 

I would say simply use `import dvc.api` instead of `from dvc import api as dvcapi`, more straightforward and almost the same length:
```python
import csv
import pickle
import dvc.api

# Loading from content
model = pickle.loads(dvc.api.read("some-model.pkl", repo="https://github.com/..."))

# Loading using file descriptor
with dvc.api.open("dataset.csv", repo=...) as fd:
    reader = csv.reader(fd)
    for row in reader:
        # ...    

# Obtaining an url
resource_url = dvc.api.get_url("path/to/resource.ext", repo=..., remote="s3")
```

**Comment by Suor** [comment]
@Naba7 I would start with some **Usage** section, with short and most common section, the continue with complete API listing.

Or another layout: Install, Usage, Methods sections. Then each method goes on its separate page, with full operation and params description, more examples. The point is making it glanceable and copy-pastable.

**Comment by Naba7** [comment]
I think the layout : Install, Usage, Methods and describing each methods is better. 
@shcheklein @jorgeorpinel If you agree to this, I will start working on the same.

**Comment by shcheklein** [comment]
@Naba7 yep, I like the idea. So, we can start with three levels:

`Python API` is the top most

it includes `Install`, `Usage`, `Method Reference` 

`Method Reference` includes one page per each method with simple example. And we need to discuss the structure for it.

@jorgeorpinel any thoughts on this?

**Comment by jorgeorpinel** [comment]
Agree. Looking forward to see a first version PR 🙂

**Comment by Naba7** [comment]
I am sorry. I won't be able to work further on this PR.

**Comment by shcheklein** [comment]
@Naba7 np! thank you for all your contributions ;)

**Comment by jorgeorpinel** [comment]
> They have some short docstrings, I will update them based on future docs or discussion here if we decide to do that.

@Suor please see iterative/dvc/issues/3092

**Comment by jorgeorpinel** [comment]
> @Suor are repo=None, rev=None, remote=None default values actually None? Or do they get turned into repo='.', rev='HEAD', remote=(read from config file)?

@Suor can you confirm about this Q ^ please? Thanks
